aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,467 @@
+  scalar_t* values_data_base = values.data_ptr<scalar_t>();
+  iscalar_t* indices_data_base = indices.data_ptr<iscalar_t>();
+  scalar_t* out_values_data_base = out_values.data_ptr<scalar_t>();
+  auto sizes = input.sizes().vec();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  /* Compute pool indices */
+  std::vector<iscalar_t> offsets = get_offsets<iscalar_t>(indices, sizes, dim);
+  iscalar_t mx_p;
+  std::vector<iscalar_t> pool = get_dense_offsets<iscalar_t>(mx_p, offsets);
+
+  /* Compute mx - a max tensor along sparse dimension */
+  std::vector<int64_t> mx_sizes(sizes.begin() + sparse_dim - 1, sizes.end());
+  mx_sizes[0] = mx_p;
+  at::Tensor mx = at::empty(mx_sizes, values.options());
+  scalar_t* mx_data_base = mx.data_ptr<scalar_t>();
+  {
+    auto ninf = -std::numeric_limits<scalar_t>::infinity();
+    for (int64_t j=0; j < nvalues * mx_p; j++) {
+      mx_data_base[j] = ninf;
+    }
+  }
+
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    scalar_t* values_data = values_data_base + i * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      mx_data[j] = std::max(mx_data[j], values_data[j]);
+    }
+  }
+
+  /* apply exp to (v - mx) and sum the results */
+  at::Tensor exp_sums = at::zeros_like(mx);
+  scalar_t* exp_sums_data_base = exp_sums.data_ptr<scalar_t>();
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    scalar_t* values_data = values_data_base + i * nvalues;
+    scalar_t* out_values_data = out_values_data_base + i * nvalues;
+    scalar_t* exp_sums_data = exp_sums_data_base + p * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      auto v = std::exp(values_data[j] - mx_data[j]);
+      if (!LogSoftMax) {
+        out_values_data[j] = v;
+      }
+      exp_sums_data[j] += v;
+    }
+  }
+
+  for (int64_t j=0; j < nvalues * mx_p; j++) {
+    if (LogSoftMax) {
+      mx_data_base[j] += std::log(exp_sums_data_base[j]);
+    } else {
+      exp_sums_data_base[j] = 1.0 / exp_sums_data_base[j];
+    }
+  }
+
+  /* normalize with the sum of exponents */
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* values_data = values_data_base + i * nvalues;
+    scalar_t* out_values_data = out_values_data_base + i * nvalues;
+    scalar_t* exp_sums_data = exp_sums_data_base + p * nvalues;
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      if (LogSoftMax) {
+        out_values_data[j] = values_data[j] - mx_data[j];
+      } else {
+        out_values_data[j] *= exp_sums_data[j];
+      }
+    }
+  }
+
+}
+
+template <typename scalar_t, typename iscalar_t, bool LogSoftMax>
+void cpu_sparse_coo_softmax_backward(Tensor& grad_input, const Tensor& grad, const Tensor& output, const int64_t dim) {
+  /*
+    gI_i = sum_j d<output_i>/d<input_j> * grad_j = sum_j output_i * (1[i==j] - output_j) * grad_j
```

Comment:
`gI_i` = sum_j d<output_j>/d<input_i>`

Reply:
And the formula is wrt to LogSoftMax=false in here.

Reply:
And the formula below is for LogSoftMax=false.

Reply:
Right, fixed. Thanks!